### PR TITLE
chore: disable reduction in LLVM eDSL

### PR DESCRIPTION
**Pull request by tobiasgrosser**
This change is needed to update to lean 4.9. We commit this as a separate change, such that for future debugging sessions there is a commit where easily both versions can be inspected.

### CODE REVIEW

**Comment by bollu** [comment]
I'm happy to enable this once the CI goes green.

File: SSA/Projects/InstCombine/LLVM/EDSL.lean [hunk 1]
```diff
@@ -249,7 +249,6 @@ elab "[alive_icom (" mvars:term,* ")| " reg:mlir_region "]" : term => do
     if ssa.alive_icom_reduce.get (← getOptions)
     then reduce com
```
Comment (bollu) [inline SSA/Projects/InstCombine/LLVM/EDSL.lean:250]:
```suggestion
    then do 
      logWarningAt reg "simplifying proof state with reduction is deprecated, as it unstable across Lean versions. Recall that reduction in Lean is slow and heuristic based, and the heuristics can change across Lean versions. Suggestion: Find appropriate simp lemmas to simplify the proof state."
      reduce com
```

**Comment by alexkeizer** [comment]
I think this is fine to merge, but I'm a bit scared of the implications. Presumably our use of `reduce` in the AST -> ICom transformations is similarly fragile. However, if we did the same there that means we'd have to think about the appropriate simp sets to reduce the `TransformDialect` machinery for every new dialect. This changes our story from "You only have to implement a few typeclasses to get pretty syntax" to "You have to implement a few typeclasses, and then spent a non-trivial amount of time engineering a simp-set to ensure it reduces properly".

For now, it's fine to just let the `LLVM` dialect be a bit of a special case, but if we start running into issues with the other `reduce`, I would strongly suggest we revisit the decision to do all of this in object code. If meta variables and AST transformations were happening on the level of Lean `Expr`s, we wouldn't have to deal with any of these simp-sets or reduction problems (and I strongly suspect it'd be significantly faster, too)

File: SSA/Projects/InstCombine/ForLean.lean [hunk 2]
```diff
@@ -84,7 +84,12 @@ lemma one_shiftLeft_mul_eq_shiftLeft {A B : BitVec w} :
   simp only [bv_toNat, Nat.mod_mul_mod, one_mul]
   rw [Nat.mul_comm]
 
+@[simp]
 def ofInt_zero_eq (w : Nat) : BitVec.ofInt w 0 = 0#w := rfl
+
+@[simp]
+def ofInt_one_eq (w : Nat) : BitVec.ofInt w 1 = 1#w := rfl
```
Comment (alexkeizer) [inline SSA/Projects/InstCombine/ForLean.lean:91]:
```suggestion
@[simp]
lemma ofInt_one_eq (w : Nat) : BitVec.ofInt w 1 = 1#w := rfl
```
It should be `lemma` or `theorem`, but `def` is definitely wrong.

This should probably be a generic lemma about `OfNat.ofNat`, e.g., 
```lean
@[simp] lemma ofInt_ofNat (w n : Nat) : BitVec.ofInt w (OfNat.ofNat n) = n#w := by sorry
```

File: SSA/Projects/InstCombine/ForLean.lean [hunk 3]
```diff
@@ -282,7 +293,7 @@ def one_sdiv { w : Nat} {a : BitVec w} (ha0 : a ≠ 0) (ha1 : a ≠ 1)
         · rw [neg_neq_iff_neq_neg]
           rw [← negOne_eq_allOnes] at hao
           assumption
-      · simp [BitVec.ofInt_zero_eq]
+      ·
         apply BitVec.udiv_one_eq_zero
```
Comment (alexkeizer) [inline SSA/Projects/InstCombine/ForLean.lean:297]: whitespace
```suggestion
      · apply BitVec.udiv_one_eq_zero
```

File: SSA/Projects/InstCombine/ForLean.lean (hunk 2 above)
Comment (tobiasgrosser) [reply SSA/Projects/InstCombine/ForLean.lean:91]: I dropped this change.

**Pull request merged**